fix backtracking

xiyuoh opened the pull request:
This PR fixes the robots' backtracking behaviour by:
- Adding another layer of waypoint filtering by removing the first waypoint if the robot's current position is closer to the second waypoint, i.e. prevent robot from backtracking to the first waypoint if it is already on the way to the second
- Modifying RobotCommandHandle's `next_arrival_estimator` following [estimation.cpp](https://github.com/open-rmf/rmf_ros2/blob/main/rmf_fleet_adapter/src/rmf_fleet_adapter/estimation.cpp#L95-L107)

(+ minor fix: shifted waypoint in clinic map slightly to open right door from inside nurse center)

The behaviour is common in hotel and clinic world. To reproduce:
``` bash
# after launching hotel world
ros2 run rmf_demos_tasks dispatch_loop -s deliverybot_charger -f kitchen -n 3 --use_sim_time

# after launching clinic world
ros2 run rmf_demos_tasks dispatch_loop -s L1_left_nurse_center -f L1_right_nurse_center -n 5 --use_sim_time
```

Example of backtracking:

https://user-images.githubusercontent.com/30265743/150392460-114e4fa9-bd1e-452a-bc82-56ae5ecc5df8.mp4

Yadunund commented on rmf_demos_fleet_adapter/rmf_demos_fleet_adapter/RobotCommandHandle.py:
```diff
@@ -550,6 +549,14 @@ def filter_waypoints(self, wps: list, threshold=1.0):
+            first_to_next = \
+                self.dist(waypoints[0].position, waypoints[1].position)
+            if dist_to_next < first_to_next:
+                del waypoints[0]
```
We do not want to delete any waypoints, especially the first one as the fleet adapter could have been interrupted with a new path where the robot is expected to wait exactly where it was to avoid conflicts with other robots. So the expectation here is that the robot will need to wait at that first waypoint until a certain time has passed before proceeding to the next. But with this implementation, the robot will go to the second waypoint without waiting.

xiyuoh commented:
Thanks for pointing out that waypoints should not be deleted, didn't notice the consequences before.

Instead, I'm shifting the first waypoint slightly nearer to the second waypoint under specific circumstances. So the `filter_waypoints` function doesn't always/unnecessarily move the first waypoint, only when it feels that the robot's current position is slightly closer to the second waypoint than the first.

Tested this with some conflict cases such as `office_conflict.launch.xml`, seems to work ok so far. The thresholds used inside the function are determined experimentally, so they may not completely eliminate backtracking, but I would say there is a significant improvement.

xiyuoh commented:
Updates to the fix:
- Added a stop callback whenever interrupt is triggered and decreased backtrack threshold as you suggested, reduces backtracking occurrences
- Modified `stop()` endpoint in `fleet_manager` as previously it was sending empty PathRequest to slotcar, but slotcar [invalidates these requests](https://github.com/open-rmf/rmf_simulation/blob/main/rmf_robot_sim_common/src/slotcar_common.cpp#L170-L176), so `stop()` wasn't doing anything at all. Changed it to a PauseRequest instead that resumes after 0.1s (initially wanted to resume only when a new PathRequest is sent, but the pause is too long for office conflict and it doesn't get resolved)
- Added a filter for spinning in front of doors, especially clinic. This started to show up after implementing the stop callback in `trigger_interrupt`, possibly because robots [do not stop turning during pauses](https://github.com/open-rmf/rmf_simulation/blob/main/rmf_robot_sim_common/src/slotcar_common.cpp#L670-L674). May not work as well at higher RTF (>2)

Requires open-rmf/rmf_ros2#167 for stop callback

Yadunund commented:
@xiyuoh is this ready for a review or do we have some more changes we want to make?

xiyuoh commented:
Ready for review:
- Changed stop endpoint to PathRequest to current position instead of PauseRequest
- Removed stop callback (no longer require the rmf_ros2 PR)
- Removed all filters except for pruning the first waypoint if it is within threshold - take it as the robot is already there

With updates from the big merges, the backtracking issue is almost non-existent in single robot loops. Still occurs sometimes with multiple robots, which is why i left the pruning in.

Yadunund commented on rmf_demos_fleet_adapter/rmf_demos_fleet_adapter/RobotCommandHandle.py:
```diff
@@ -542,78 +540,21 @@ def filter_waypoints(self, wps: list, threshold=1.0):
-            else:
-                index = index + 1
+                self.dist(first_position, last_pose) < threshold and\
+                abs(first_position[2] - last_pose[2]) < 0.1:
```
What is the rationale for checking this difference in `yaw`?

xiyuoh replied:
That was initially added to prevent unnecessary spinning in front of doors when the robot gets interrupted, just tested and seems like the spinning only occurs at higher RTF (> 2.0) with or without checking yaw. Removed in last commit

Yadunund commented on rmf_demos_fleet_adapter/rmf_demos_fleet_adapter/fleet_manager.py:
```diff
@@ -18,6 +18,7 @@
 import math
 import yaml
 import argparse
+import time
```
`time` seems to be unused

xiyuoh replied:
Removed!

Yadunund reviewed (changes requested):
Looks good. Thanks for iterating. Minor comments to address

Yadunund commented on rmf_demos_fleet_adapter/rmf_demos_fleet_adapter/RobotCommandHandle.py:
```diff
@@ -189,6 +182,12 @@ def follow_new_path(
             path_finished_callback):
 
         self.stop()
```
With https://github.com/open-rmf/rmf_ros2/pull/167, do we still need to call `self.stop()` here?

xiyuoh replied:
You're right we don't, especially with demos' stop endpoint. Updated accordingly

Yadunund reviewed (approved):
LGTM
